pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,373 @@
+for mesh metadata. It provides bounded degree and amplification factor
+with the meshsub construction and augments it using gossip propagation
+of metadata with the randomsub technique.
+
+
+## The gossipsub protocol
+
+We can now provide a specification of the pubsub protocol by sketching
+out the router implementation.  The router is backwards compatible
+with floodsub, as it accepts floodsub peers and behaves like floodsub
+towards them.
+
+### Control messages
+
+The protocol defines four control messages:
+- `GRAFT`: graft a mesh link; notifies the peer that it has been added to the local mesh view.
+- `PRUNE`: prune a mesh link; notifies the peer that it has been removed from the local mesh view.
+- `IHAVE`: gossip; notifies the peer that the following messages were recently seen and are available on request.
+- `IWANT`: requests the out of band transmission of messages announced in an `IHAVE` message.
+
+### Router state
+
+The router maintains the following state:
+- `peers`: a set of all known peers; `peers.gossipsub` denotes the gossipsub peers
+   while `peers.floodsub` denotes the floodsub peers.
+- `mesh`: the overlay meshes as a map of topic to lists of peers.
+- `fanout`: the mesh peers to which we are publishing to without topic membership,
+   as a map of topic to list of peers.
+- `seen`: this is the timed message id cache, which tracks seen messages.
+- `mcache`: a message cache that contains the messages for the last few
+   heartbeat ticks.
+
+The message cache is a data structure that stores windows of message ids
+and the corresponding messages. It supports the following operations:
+- `mcache.put(m)`: adds a message to the current window and the cache.
+- `mcache.get(id)`: retrieves a message from the cache by its id, if it is still present.
+- `mcache.window()`: retrieves the message id for messages in the current histor window.
+- `mcache.shift()`: shifts the current window, discarding messages older than the
+   history length of the cache.
+
+The timed message id cache is the flow control mechanism. It tracks
+the message ids of seen message for the last couple of minutes. It is
+separate from `mcache` for implementation reasons in Go (the `seen`
+cache is inherited from the pubsub framework), but they could be the
+same data structure.
+
+### Topic membership
+
+Topic membership is controlled by two operations supported by the
+router, as part of the pubsub api:
+- On `JOIN(topic)` the router joins the topic. In order to do so, it
+  selects `D` peers from `peers.gossipsub[topic]`, adds them to `mesh[topic]`
+  and notifies them with a `GRAFT(topic)` control message. If it already has
+  `fanout` peers in the topic, then it selects those peers as the
+  initial mesh peers.
+- On `LEAVE(topic)` the router leaves the topic. It notifies the peers in
+  `mesh[topic]` with a `PRUNE(topic)` message and forgets `mesh[topic]`.
+
+Note that the router can publish messages without topic membership. In order
```

Comment:
Is there a relevant application for publishing messages one is not subscribed to? To me it seems like this makes the protocol more complicated for very little gain.

To simplify it while keeping the option of publishing without membership, one could just use "randomsub" which would be equivalent as long as the time between those messages is greater than the fanout TTL.

Reply:
Well, we still want to keep stable routes for non-subscribing publishers, hence the fanout set.

Re: applications: any streaming application will have this property, the source doesn't need to subscribe at all.

Reply:
Another application is when you have a cluster utilizing pubsub and allow clients to publish to the cluster by using pubsub. The clients are not interested in the pubsub chatter in that case.